src/main/java/com/gluonhq/richtext/RichTextAreaSkin.java

```diff
@@ -156,19 +161,52 @@ public void dispose() {
 
     /// PRIVATE METHODS /////////////////////////////////////////////////////////
 
-    //TODO Need more optimal way of rendering text fragments.
+    // TODO Need more optimal way of rendering text fragments.
     //  For now rebuilding the whole text flow
     private void refreshTextFlow() {
         fontCacheEvictionTimer.pause();
         try {
             var fragments = new ArrayList<Text>();
-            viewModel.walkFragments((text, decoration) -> fragments.add(buildText(text, decoration)));
+            var backgroundIndexRanges = new ArrayList<IndexRangeColor>();
+            var length = new AtomicInteger();
+            viewModel.walkFragments((text, decoration) -> {
+                final Text textNode = buildText(text, decoration);
+                fragments.add(textNode);
+
+                if (decoration.getBackground() != Color.TRANSPARENT) {
+                    final IndexRangeColor indexRangeColor = new IndexRangeColor(
+                            length.get(),
+                            length.get() + textNode.getText().length(),
+                            decoration.getBackground()
+                    );
+                    backgroundIndexRanges.add(indexRangeColor);
+                }
+                length.addAndGet(textNode.getText().length());
+            });
             textFlow.getChildren().setAll(fragments);
+            addBackgroundPathsToLayers(backgroundIndexRanges);
         } finally {
             fontCacheEvictionTimer.start();
         }
     }
 
+    private void addBackgroundPathsToLayers(List<IndexRangeColor> backgroundIndexRanges) {
+        if (layers.getChildren().size() > 3) {
+            layers.getChildren().removeIf(node -> "background".equals(node.getId()));
+        }
+        final List<Path> backgroundPaths = backgroundIndexRanges.stream()
+                .map(indexRangeBackground -> {
+                    final Path path = new Path(textFlow.rangeShape(indexRangeBackground.getStart(), indexRangeBackground.getEnd()));
```

Comment:
The PR looks good, but I'm concern about performance of removing and adding nodes all over again on every refresh.
Is there a way we could cache these paths and invalidate them only when something changed?

Reply:
I have used an `ObservableSet` to cache these Paths and only add them to the `Group` if they didn't exist.